graph.py

```diff
@@ -74,9 +110,62 @@ def _remove_switch_hops(circuit):
     def shortest_paths(self, source, destination, parameter=None):
         """Calculate the shortest paths and return them."""
         try:
-            paths = list(nx.shortest_simple_paths(self.graph,
-                                                  source,
-                                                  destination, parameter))
+            paths = list(self._path_fun(self.graph,
+                                        source, destination, parameter))
         except (NodeNotFound, NetworkXNoPath):
             return []
         return paths
+
+    def constrained_flexible_paths(self, source, destination,
+                                   depth=None, **metrics):
+        """Calculate the constrained shortest paths with flexibility."""
+        base = metrics.get("base", {})
+        flexible = metrics.get("flexible", {})
+        # Retrive subgraph with edges that meet base requirements.
+        default_edge_list = list(self._filter_edges(
+            self.graph.edges(data=True), **base))
+        length = len(flexible)
+        if depth is None:
+            depth = length
+        depth = min(length, max(0, depth))
+        results = []
+        paths = []
+        i = 0
+        # Create "sub-subgraphs" from original subgraph by trimming edges
+        # that fail to meet flexible requirements. Search for a shortest
+        # path in each of these graphs, until at least one is found.
+        while (paths == [] and i in range(0, depth+1)):
+            for combo in combinations(flexible.items(), length-i):
+                additional = dict(combo)
+                paths = self._constrained_shortest_paths(
+                    source, destination, ((u, v) for u, v, d in
+                                          self._filter_edges(default_edge_list,
+                                                             **additional)))
+                if paths != []:
+                    results.append(
+                        {"paths": paths, "metrics": {**base, **additional}})
+            i = i + 1
```

Comment:
This loop runs in <img src="https://render.githubusercontent.com/render/math?math=O(2^n)" />. Have you stressed this method using a graph with a high number of edges to see if it runs in a reasonable time?

Reply:
Indeed, test_search_2 uses a graph with 81 edges and runs reasonably quickly, but it could be faster.

I checked the code and found that it is indeed **O(2^n)** but the number of edges does not affect the number of computations. Only the size of the flexible metrics dictionary. I will double-check though.

Reply:
Yes, you're right, the combinations are on the metrics, not the edges. My mistake.